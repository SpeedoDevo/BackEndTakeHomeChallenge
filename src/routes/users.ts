import UsersService from '../services/UsersService'
import BadRequestError from '../errors/BadRequestError'
import IUser from '../models/IUser'
import Router from 'koa-router';

const usersService = new UsersService()
const router = new Router()

router.get('/users', async ctx => {
  const users: Array<IUser> = usersService.getAll()

  ctx.body = {
    users: users.map(user => user.toObject())
  }
})

router.post('/users', async ctx => {
  const { name, username } = ctx.request.body
  if (!name) {
    throw new BadRequestError('Name not defined.')
  }

  const user: IUser = usersService.create(name, username)

  ctx.body = {
    user: user.toObject()
  }
})

export default router.routes()
