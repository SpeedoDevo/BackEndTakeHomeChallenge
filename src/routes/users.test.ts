import { createMockContext } from '@shopify/jest-koa-mocks'
import { Context } from 'koa'
import { default as _usersRoutes } from './users'

const usersRoutes = _usersRoutes as (ctx: Context) => Promise<void>

describe('users routes', () => {
  let ctx: Context

  describe('GET /users', () => {
    describe('without query', () => {
      beforeEach(async () => {
        ctx = createMockContext({ method: 'GET', url: '/users' })

        await usersRoutes(ctx)
      })

      it('should send the right response', () => {
        expect(ctx.body).toMatchSnapshot()
      })
    })

    describe('with query', () => {
      beforeEach(async () => {
        ctx = createMockContext({ method: 'GET', url: '/users?q=Stark' })

        await usersRoutes(ctx)
      })

      it('should send the right response', () => {
        expect(ctx.body).toMatchSnapshot()
      })
    })
  })

  describe('POST /users', () => {
    beforeAll(async () => {
      ctx = createMockContext({
        method: 'POST',
        url: '/users',
        requestBody: { name: 'new user', username: 'admin' },
      })

      await usersRoutes(ctx)
    })

    it('should send the right response', () => {
      expect(ctx.body).toMatchSnapshot()
    })

    describe('after creating', () => {
      beforeEach(async () => {
        ctx = createMockContext({ method: 'GET', url: '/users' })

        await usersRoutes(ctx)
      })

      it('should send the right response', () => {
        expect(ctx.body).toMatchSnapshot()
      })
    })
  })

  describe('PUT /users', () => {
    beforeAll(async () => {
      ctx = createMockContext({
        method: 'PUT',
        url: '/users',
        requestBody: {
          id: 1,
          name: 'Daenerys',
          username: 'thequeenbee',
          avatar: 'some dragon or something',
        },
      })

      await usersRoutes(ctx)
    })

    it('should send the right response', () => {
      expect(ctx.body).toMatchSnapshot()
    })

    describe('after updating', () => {
      beforeEach(async () => {
        ctx = createMockContext({ method: 'GET', url: '/users' })

        await usersRoutes(ctx)
      })

      it('should send the right response', () => {
        expect(ctx.body).toMatchSnapshot()
      })
    })
  })
})
