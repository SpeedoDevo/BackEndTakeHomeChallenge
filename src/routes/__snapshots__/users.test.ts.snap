// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users routes GET /users with query should send the right response 1`] = `
Object {
  "users": Array [
    Object {
      "avatar": undefined,
      "id": 3,
      "name": "Sansa Stark",
      "username": "joffery.fan",
    },
    Object {
      "avatar": undefined,
      "id": 4,
      "name": "Ned Stark",
      "username": "honouraboveall",
    },
  ],
}
`;

exports[`users routes GET /users without query should send the right response 1`] = `
Object {
  "users": Array [
    Object {
      "avatar": undefined,
      "id": 1,
      "name": "Daenerys Targaryen",
      "username": "thequeenbee",
    },
    Object {
      "avatar": undefined,
      "id": 2,
      "name": "Jon Snow",
      "username": "forthebrotherhood",
    },
    Object {
      "avatar": undefined,
      "id": 3,
      "name": "Sansa Stark",
      "username": "joffery.fan",
    },
    Object {
      "avatar": undefined,
      "id": 4,
      "name": "Ned Stark",
      "username": "honouraboveall",
    },
    Object {
      "avatar": undefined,
      "id": 5,
      "name": "Cersei Lannister",
      "username": "bestmuminwesteros",
    },
  ],
}
`;

exports[`users routes POST /users after creating should send the right response 1`] = `
Object {
  "users": Array [
    Object {
      "avatar": undefined,
      "id": 1,
      "name": "Daenerys Targaryen",
      "username": "thequeenbee",
    },
    Object {
      "avatar": undefined,
      "id": 2,
      "name": "Jon Snow",
      "username": "forthebrotherhood",
    },
    Object {
      "avatar": undefined,
      "id": 3,
      "name": "Sansa Stark",
      "username": "joffery.fan",
    },
    Object {
      "avatar": undefined,
      "id": 4,
      "name": "Ned Stark",
      "username": "honouraboveall",
    },
    Object {
      "avatar": undefined,
      "id": 5,
      "name": "Cersei Lannister",
      "username": "bestmuminwesteros",
    },
    Object {
      "avatar": undefined,
      "id": 6,
      "name": "new user",
      "username": "admin",
    },
  ],
}
`;

exports[`users routes POST /users should send the right response 1`] = `
Object {
  "user": Object {
    "avatar": undefined,
    "id": 6,
    "name": "new user",
    "username": "admin",
  },
}
`;

exports[`users routes PUT /users after updating should send the right response 1`] = `
Object {
  "users": Array [
    Object {
      "avatar": "some dragon or something",
      "id": 1,
      "name": "Daenerys",
      "username": "thequeenbee",
    },
    Object {
      "avatar": undefined,
      "id": 2,
      "name": "Jon Snow",
      "username": "forthebrotherhood",
    },
    Object {
      "avatar": undefined,
      "id": 3,
      "name": "Sansa Stark",
      "username": "joffery.fan",
    },
    Object {
      "avatar": undefined,
      "id": 4,
      "name": "Ned Stark",
      "username": "honouraboveall",
    },
    Object {
      "avatar": undefined,
      "id": 5,
      "name": "Cersei Lannister",
      "username": "bestmuminwesteros",
    },
    Object {
      "avatar": undefined,
      "id": 6,
      "name": "new user",
      "username": "admin",
    },
  ],
}
`;

exports[`users routes PUT /users should send the right response 1`] = `
Object {
  "user": Object {
    "avatar": "some dragon or something",
    "id": 1,
    "name": "Daenerys",
    "username": "thequeenbee",
  },
}
`;
